src/passes/stringify-walker.h

```diff
@@ -75,4 +76,98 @@ struct StringifyWalker
 
 #include "stringify-walker-impl.h"
 
+namespace wasm {
+
+/*
+ * This custom hasher conforms to std::hash<Key>. It's purpose is to provide
+ * a custom hash for if expressions, so the if-condition of the if expression is
+ * not included in the hash for the if expression. This is needed because in the
+ * binary format, the if-condition comes before and is consumed by the if. To
+ * evaluate in the same order as the binary format, we hash the if condition
+ * before and separately from the rest of the if expression.
+ */
+struct StringifyHasher {
+  size_t operator()(Expression* curr) const {
+    if (Properties::isControlFlowStructure(curr)) {
+      if (auto* iff = curr->dynCast<If>()) {
+        return hashIfNoCondition(iff);
+      }
+
+      return ExpressionAnalyzer::hash(curr);
+    }
+
+    return ExpressionAnalyzer::shallowHash(curr);
+  }
+
+  static uint64_t hashIfNoCondition(If* iff) {
+    size_t digest = wasm::hash(0);
+    rehash(digest, iff->_id);
+    rehash(digest, iff->type.getID());
+    rehash(digest, ExpressionAnalyzer::hash(iff->ifTrue));
+    rehash(digest, iff->ifTrue->_id);
+    rehash(digest, iff->ifTrue->type.getID());
+
+    if (iff->ifFalse) {
+      rehash(digest, ExpressionAnalyzer::hash(iff->ifFalse));
+      rehash(digest, iff->ifFalse->_id);
+      rehash(digest, iff->ifFalse->type.getID());
+    }
+
+    return digest;
+  }
+};
+
+/*
+ * This custom equator conforms to std::equal_to<Key>. Similar to
+ * StringifyHasher, it's purpose is to not include the if-condition when
+ * evaluating the equality of two if expressions.
+ */
+struct StringifyEquator {
+  bool operator()(Expression* lhs, Expression* rhs) const {
+    if (Properties::isControlFlowStructure(lhs) &&
+        Properties::isControlFlowStructure(rhs)) {
+      auto* iffl = lhs->dynCast<If>();
+      auto* iffr = rhs->dynCast<If>();
+
+      if (iffl && iffr) {
+        return equalIfNoCondition(iffl, iffr);
+      }
+
+      return ExpressionAnalyzer::equal(lhs, rhs);
+    }
+
+    return ExpressionAnalyzer::shallowEqual(lhs, rhs);
+  }
+
+  bool equalIfNoCondition(If* iffl, If* iffr) const {
+    return ExpressionAnalyzer::equal(iffl->ifTrue, iffr->ifTrue) &&
+           ExpressionAnalyzer::equal(iffl->ifFalse, iffr->ifFalse);
+  }
+};
+
+struct HashStringifyWalker : public StringifyWalker<HashStringifyWalker> {
+  /*
+   * After calling walkModule, this vector contains the result of encoding a
+   * wasm module as a string. Each value represents either an Expression or a
+   * separator to mark the end of control flow.
+   */
+  std::vector<uint64_t> hashString;
+  uint64_t monotonic = 0;
+  /*
+   * Contains a mapping of expression pointer to monotonic value to ensure we
+   * use the same monotonic value for matching expressions. A custom hasher and
```

Comment:
We can probably drop "monotonic" everywhere in this comment without losing any useful information.

Reply:
done